packages/utils/src/metrics.ts

```diff
@@ -0,0 +1,80 @@
+export type LabelsGeneric = Record<string, string | number>;
+export type LabelKeys<Labels extends LabelsGeneric> = Extract<keyof Labels, string>;
+export type CollectFn<Labels extends LabelsGeneric> = (metric: Gauge<Labels>) => void;
+
+// Type `{}` is required here as alternative types do not work
+// eslint-disable-next-line @typescript-eslint/ban-types
+export type NoLabels = {};
+
+export interface Gauge<Labels extends LabelsGeneric = NoLabels> {
+  startTimer(): NoLabels extends Labels ? () => number : (labels: Labels) => number;
+  startTimer(labels: NoLabels extends Labels ? undefined : Labels): (labels?: Labels) => number;
+
+  inc: NoLabels extends Labels ? (value?: number) => void : (labels: Labels, value?: number) => void;
```

Comment:
> do we have to increment every label the same time, else this type will raise error.

I am not sure what you mean by "increment every label at the same time". What this does is that if you metric has a label defined, then when setting a value for that metric it must specify a label.

Let's look at an example

```ts
requestErrors: register.gauge<{routeId: string}>({
  name: "vc_rest_api_client_request_errors_total",
  help: "Total count of errors on REST API client requests by routeId",
  labelNames: ["routeId"],
}),
```

This is no longer allowed
```ts
// An argument for 'labels' was not provided.
this.metrics?.requestErrors.inc();
```
You are forced to do
```ts
this.metrics?.requestErrors.inc({routeId});
```
I don't see any reason why you would want to set this metric without a label, it would mean there is a unlabeled value which could be combined with further unlabeled values of the same metric but this defeats the purpose of adding a label in the first place, to separate tag and separate values of the same metric.

I only found a few cases were we forgot to set the label but fixed those in this PR as well.


Note that there is also still the option to make the label optional when defining the type for it, I don't think there is any reason to do this but it's an option.